### ISSUE TITLE
Document 2fa sms requirements 

### ISSUE THREAD

**Issue by karaggeorge** [octokit/octokit.js#1290]
More of a documentation request. It took me a while to figure out why I wasn't getting a 2fa code via sms when authenticating using the username/password method. After looking for a while I discovered this comment: https://github.com/sigmavirus24/github3.py/issues/387#issuecomment-110902328

The sms code is not sent until you actually try to create a token (or call `oauthAuthorizations.createAuthorization` of the API), even though the passed 2fa method gets called every time. After calling that API method, I finally get an sms, and resolving 2fa with that does the trick. 

I think it would be nice to note that somewhere in the docs

**Comment by gr2m** [comment]
I didn’t know that, thanks for sharing! I’ll give this some thought. I wonder if we could do some kind of trick to make sure that an SMS is always sent if 2FA is enabled for basic authentication 🤔

**Comment by gr2m** [comment]
I think one trick could be to actually create a token, but then delete it again right after (if 2FA is not enabled) before making the actual first request.

I also plan to add options to [`auth.*`](https://github.com/octokit/rest.js/tree/b273efe41745a8222d5093d58243298bc0b0b3e3#authentication) to let people create a token that will be used for further requests instead of using username & password, that should mitigate the problem.

**Comment by karaggeorge** [comment]
I think the docs basically say the the username/password/2fa method of login is supposed to be very short-term, specifically used to create an actual token. That's why they only send the sms when you try to create a token. With that in mind, I think the second method sounds better. So, even if they do username/password/2fa, it could generate a token (maybe also request a `note` param for the name of the token) and then use that for the rest of the calls.

I'd love to open a PR for it if you direct me to where I should make the change. Looking at the code, it looks like most of the auth happens when the first request is made, rather than when the instance is created

EDIT: I actually couldn't find anything in their docs about it, other than the comment I linked in the description which seems to be coming from their support team describing the expected behavior. GitHub should probably include that in their API documentation as well.

**Comment by gr2m** [comment]
> EDIT: I actually couldn't find anything in their docs about it, other than the comment I linked in the description which seems to be coming from their support team describing the expected behavior. GitHub should probably include that in their API documentation as well.

I’ve shared that with the docs team, they are already aware of lack of documentation around 2FA and SMS.

> I'd love to open a PR for it if you direct me to where I should make the change. Looking at the code, it looks like most of the auth happens when the first request is made, rather than when the instance is created

That would be great. Let’s discuss how the API would look like first. Can you start a pull request draft updating the `README.md` [Authentication section](https://github.com/octokit/rest.js#authentication) to add the additional options? It should probably match the `scopes` and `note` options of [`octokit.oauthAuthorizations.createAuthorization()`](https://octokit.github.io/rest.js/#api-OauthAuthorizations-createAuthorization)

**Comment by karaggeorge** [comment]
Just opened draft PR making an attempt to update the `README.md`. Not sure if we would pass more options since most of them seem to be for OAuth apps.

Also thinking about it, we should probably somehow make the token accessible to the user after creating, so they can save it for later use.

**Comment by gr2m** [comment]
> Also thinking about it, we should probably somehow make the token accessible to the user after creating, so they can save it for later use.

Good point 🤔

**Comment by HoudaDabou** [comment]
Hi there,

Any update on this issue?

Thanks in advance :)

**Comment by gr2m** [comment]
Not yet, I had to shift my priorities. I’ve described the endpoints that will trigger an SMS at https://github.com/octokit/auth.js#basic-and-personal-access-token-authentication, hope that helps for the time being

**Comment by HoudaDabou** [comment]
yes, thanks for the quick reply!

**Comment by gr2m** [comment]
I know this took a long while, but this issue will finally be resolved proper by #1550. You will be able to do this:

```js
const { createBasicAuth } = require('@octokit/auth-basic')
const octokit = new Octokit({
  authStrategy: createBasicAuth
  auth: {
    username,
    password,
    on2Fa
  }
})
```

This will create the authentication token internally and enforce that an SMS gets sent in any case.
